#' Pseudobulk analysis of allele frequencies
#'
#' Perform pseudobulk analysis to identify most common alleles or indel patterns in data while ignoring cell barcodes. Only used for CRISPR-Cas9 for now.
#'
#' @param crispresso_allele_df Cleaned allele dataframe generated by `clean_crispresso_output()`
#' @param ignore_subs Ignore substitutions since CRISPR-Cas9 only produces indels. Set to FALSE if running on base/prime editing data. Default is TRUE.
#' @returns A dataframe with alleles and their total read counts
#' @examples
#' data(cocktail)
#' mydata <- clean_crispresso_output(cocktail)
#' mytab <- analyze_alleles(mydata)
#'
analyze_alleles <- function(crispresso_allele_df, ignore_subs = TRUE) {

  output_df <- crispresso_allele_df %>%
    janitor::clean_names()

  if (ignore_subs) {

    output_df <- output_df %>%
      dplyr::mutate(indel_profile_aligned = gsub("A|T|C|G", "N", aligned_sequence),
                    indel_profile_reference = gsub("A|T|C|G", "N", reference_sequence),
                    edited_allele_pattern = paste(indel_profile_reference, indel_profile_aligned, sep = " | ")) %>%
      dplyr::group_by(indel_profile_aligned, indel_profile_reference, edited_allele_pattern) %>%
      dplyr::summarize(number_reads = sum(number_reads)) %>%
      dplyr::ungroup() %>%
      dplyr::arrange(desc(number_reads)) %>%
      dplyr::mutate(percent_reads = number_reads / sum(number_reads) * 100,
                    number_deletions = stringr::str_count(as.character(indel_profile_aligned), "\\-"),
                    number_insertions = stringr::str_count(as.character(indel_profile_reference), "\\-"))


  } else {

    output_df <- output_df %>%
      dplyr::mutate(edited_allele_pattern = paste(reference_sequence, aligned_sequence, sep = " | ")) %>%
      dplyr::arrange(desc(number_reads)) %>%
      dplyr::mutate(percent_reads = number_reads / sum(number_reads) * 100,
                    number_deletions = stringr::str_count(as.character(aligned_sequence), "\\-"),
                    number_insertions = stringr::str_count(as.character(reference_sequence), "\\-")) %>%
      dplyr::group_by(aligned_sequence, reference_sequence, edited_allele_pattern) %>%
      dplyr::summarize(number_reads = sum(number_reads)) %>%
      dplyr::ungroup() %>%
      dplyr::arrange(desc(number_reads)) %>%
      dplyr::mutate(percent_reads = number_reads / sum(number_reads) * 100,
                    number_deletions = stringr::str_count(as.character(aligned_sequence), "\\-"),
                    number_insertions = stringr::str_count(as.character(reference_sequence), "\\-"))

  }

  return(output_df)

}

#' Clean CRISPResso output
#'
#' Prepare cell allele frequency dataframe for genotyping analysis
#'
#' @param crispresso_allele_df Dataframe containing allele read counts for each cell barcode. Generated by running CRISPResso on each cell barcode fastq file.
#' @param allele_read_coverage_threshold Integer value indicating the limit of detection for rare indels. Alleles/indels with read counts less than this value are assumed to be sequencing errors and will be removed from data. Default is 2.
#' @param ignore_subs Ignore substitutions since CRISPR-Cas9 only produces indels. Set to FALSE if running on base/prime editing data. Default is TRUE.
#' @param af_cutoff Integer corresponding to minimum pseudobulk percent that an allele must surpass in order to be analyzed downstream. Default is 0.
#' @param target_id Name of loci being targeted; typically name/ID of gRNA used when running CRISPResso.
#' @param update_unmodified_allele_sequence Force aligned allele sequence to match reference allele if it is called as "Unedited" by CRISPResso2. Defaults to TRUE.
#' @returns A dataframe with cell barcodes, allele read counts, and allele sequences used for downstream genotyping analysis.
#' @examples
#' data(cocktail)
#' mydata <- clean_crispresso_output(cocktail)
#'
clean_crispresso_output <- function(crispresso_allele_df,
                                    allele_read_coverage_threshold = 2,
                                    ignore_subs = TRUE,
                                    af_cutoff = 0,
                                    target_id = NULL,
                                    update_unmodified_allele_sequence = TRUE) {

  slim_df <- crispresso_allele_df %>%
    janitor::clean_names()

  if (update_unmodified_allele_sequence) {

    wt_seq <- analyze_alleles(crispresso_allele_df = crispresso_allele_df, ignore_subs = FALSE) %>%
      dplyr::filter(aligned_sequence == reference_sequence) %>%
      dplyr::pull(reference_sequence)
    wt_seq <- wt_seq[!grepl("-", wt_seq)]
    stopifnot(length(wt_seq) == 1)

    slim_df <- slim_df %>%
      dplyr::mutate(aligned_sequence = ifelse(unedited, wt_seq, aligned_sequence),
                    reference_sequence = ifelse(unedited, wt_seq, reference_sequence))

  }

  if (!tibble::is_tibble(slim_df)) {
    slim_df <- slim_df %>% dplyr::rename(number_reads = x_reads)
  }

  slim_df <- slim_df %>%
    dplyr::select(sample, target_id, reference_sequence, aligned_sequence, unedited, number_reads) %>%
    dplyr::filter(target_id %in% target_id)

  if (ignore_subs) {

    slim_df <- slim_df %>%
      dplyr::mutate(indel_profile_aligned = gsub("A|T|C|G", "N", aligned_sequence),
                    indel_profile_reference = gsub("A|T|C|G", "N", reference_sequence),
                    edited_allele_pattern = paste(indel_profile_reference, indel_profile_aligned, sep = " | ")) %>%
      dplyr::select(-reference_sequence, -aligned_sequence) %>%
      dplyr::group_by(sample, indel_profile_aligned, indel_profile_reference, edited_allele_pattern, unedited, target_id) %>%
      dplyr::summarize(number_reads = sum(number_reads)) %>%
      dplyr::ungroup()

  } else {

    slim_df <- slim_df %>%
      dplyr::mutate(edited_allele_pattern = paste(reference_sequence, aligned_sequence, sep = " | "))

  }

  top_alleles <- analyze_alleles(crispresso_allele_df = crispresso_allele_df, ignore_subs = ignore_subs)
  patterns <- top_alleles$edited_allele_pattern[top_alleles$percent_reads > af_cutoff]

  slim_df <- slim_df %>%
    dplyr::filter(number_reads > allele_read_coverage_threshold,
                  edited_allele_pattern %in% patterns)

  slim_df <- slim_df %>%
    dplyr::group_by(sample) %>%
    dplyr::mutate(percent_reads = number_reads / sum(number_reads) * 100,
                  n_alleles = n()) %>%
    dplyr::arrange(desc(percent_reads)) %>%
    dplyr::ungroup()

  return(slim_df)

}
