#' Pseudobulk analysis of allele frequencies
#'
#' Perform pseudobulk analysis to identify most common alleles or indel patterns in data while ignoring cell barcodes. Only used for CRISPR-Cas9 for now.
#'
#' @param crispresso_allele_df Cleaned allele dataframe generated by `clean_crispresso_output()`
#' @param ignore_subs Ignore substitutions since CRISPR-Cas9 only produces indels. Set to FALSE if running on base/prime editing data. Default is TRUE.
#' @returns A dataframe with alleles and their total read counts
#' @examples
#' data(cocktail)
#' mydata <- clean_crispresso_output(cocktail)
#' mytab <- analyze_alleles(mydata)
#'
get_genotype_alleles <- function(cleaned_crispresso_allele_df, hom_barcodes, het_barcodes) {

 # identify alleles associated with homozygous cells
 hom_edit_allele <- cleaned_crispresso_allele_df %>%
   dplyr::filter(sample %in% hom_barcodes) %>%
   dplyr::select(sample, edited_allele_pattern, percent_reads) %>%
   dplyr::group_by(sample) %>%
   dplyr::arrange(desc(percent_reads)) %>%
   dplyr::slice(1) %>%
   dplyr::ungroup()

 hom_edit_allele <- table(hom_edit_allele$edited_allele_pattern) %>%
   as.data.frame.table() %>%
   dplyr::arrange(desc(Freq)) %>%
   dplyr::slice(1) %>%
   dplyr::pull(Var1) %>%
   as.character()

  # identify alleles associated with heterozygous cells
 het_edit_allele <- cleaned_crispresso_allele_df %>%
   dplyr::filter(sample %in% het_barcodes) %>%
   dplyr::select(sample, edited_allele_pattern, percent_reads) %>%
   dplyr::group_by(sample) %>%
   dplyr::arrange(desc(percent_reads)) %>%
   dplyr::slice(1:2) %>%
   dplyr::ungroup()

 het_edit_allele <- table(het_edit_allele$edited_allele_pattern) %>%
   as.data.frame.table() %>%
   dplyr::arrange(desc(Freq)) %>%
   dplyr::slice(1:2) %>%
   dplyr::pull(Var1) %>%
   as.character()

  output_df <- dplyr::bind_rows(data.frame(genotype = "Hom edit", edited_allele_pattern = hom_edit_allele),
                                data.frame(genotype = "Het", edited_allele_pattern = het_edit_allele))

  return(output_df)

}

#' Summarize sample composition after genotyping cells
#'
#' Summarize cell genotype percentages after genotyping is complete
#'
#' @param cleaned_crispresso_allele_df Cleaned allele dataframe generated by `clean_crispresso_output()`.
#' @param cell_genotypes A two column dataframe with cell barcodes and their predicted genotype.
#' @param genotype_alleles_map A two column dataframe mapping each genotype category to its associated allele pattern.
#' @param multiplet_handling Determines how multiplets are handled before counting cells.
#' @returns
#' @examples
#'
#'
estimate_sample_composition <- function(cleaned_crispresso_allele_df,
                                        cell_genotypes,
                                        genotype_alleles_map,
                                        multiplet_handling = c("remove", "split"),
                                        dropout_correction = TRUE,
                                        estimated_dropout_rate = 0.123) {

  stopifnot(ncol(cell_genotypes) == 2)
  stopifnot(ncol(genotype_alleles_map) == 2)

  # make sure cell barcode genotype map contains only unique mappings
  cell_genotypes <- cell_genotypes %>% dplyr::distinct()

  comp_df <- cleaned_crispresso_allele_df %>%
    dplyr::left_join(cell_genotypes, by = "sample", relationship = "many-to-many")

  # remove multiplets before calculating sample composition
  if (multiplet_handling %in% "remove") {

    table_df <- comp_df %>%
      dplyr::select(sample, predict_genotype) %>%
      dplyr::distinct() %>%
      dplyr::filter(!predict_genotype %in% c("Transparent multiplet", "Opaque multiplet"))

    table_df <- table(table_df$predict_genotype) %>%
      as.data.frame.table() %>%
      dplyr::rename(genotype = Var1, droplet_counts = Freq) %>%
      dplyr::mutate(sample_percent = droplet_counts / sum(droplet_counts) * 100)

  # split multiplets into partial droplets based on allele frequency percentages in each droplet. this is experimental.
  } else if (multiplet_handling %in% "split") {

    multiplet_df <- comp_df %>%
      dplyr::filter(predict_genotype %in% c("Transparent multiplet", "Opaque multiplet")) %>%
      dplyr::left_join(genotype_alleles_map, by = "edited_allele_pattern", relationship = "many-to-many") %>%
      dplyr::mutate(predict_genotype = ifelse(unedited, "WT", genotype)) %>%
      dplyr::filter(!is.na(genotype)) %>% # remove alleles generated by sequencing errors
      dplyr::select(sample, edited_allele_pattern, number_reads, genotype) %>%
      dplyr::group_by(sample, genotype) %>% # combine heterozygous alleles because they're from same cell
      dplyr::summarize(number_reads = sum(number_reads), .groups = "drop") %>%
      dplyr::ungroup() %>%
      dplyr::group_by(sample) %>%
      dplyr::mutate(droplet_counts = number_reads / sum(number_reads)) %>%
      dplyr::ungroup() %>%
      dplyr::select(genotype, droplet_counts)

    singleplet_df <- comp_df %>%
      dplyr::filter(!grepl("multiplet", predict_genotype, ignore.case = TRUE)) %>%
      dplyr::select(sample, predict_genotype) %>%
      dplyr::distinct() %>%
      dplyr::mutate(droplet_counts = 1) %>%
      dplyr::rename(genotype = predict_genotype) %>%
      dplyr::select(genotype, droplet_counts)

    table_df <- dplyr::bind_rows(multiplet_df, singleplet_df) %>%
      dplyr::group_by(genotype) %>%
      dplyr::summarize(droplet_counts = sum(droplet_counts)) %>%
      dplyr::mutate(sample_percent = droplet_counts / sum(droplet_counts) * 100)

  }

  if (dropout_correction) {

    predicted_het_percent <- table_df$sample_percent[table_df$genotype %in% "Het"]
    dropout_coef <- estimated_dropout_rate * 0.01
    adjusted_het_percent <- predicted_het_percent / (1 - dropout_coef * predicted_het_percent)
    hom_adjustment_value <- (adjusted_het_percent - predicted_het_percent) / 2
    hom_adjusted_sample_percent <- table_df$sample_percent[table_df$genotype %in% c("Hom edit", "WT")] - hom_adjustment_value

    if (min(hom_adjusted_sample_percent) < 0) {
      hom_adjusted_sample_percent <- hom_adjusted_sample_percent + min(hom_adjusted_sample_percent)
      hom_adjusted_sample_percent <- unlist(lapply(hom_adjusted_sample_percent, function(arg) max(c(arg, 0))))
    }

    table_df <- table_df %>%
      dplyr::mutate(adjusted_sample_percent = c(adjusted_het_percent, hom_adjusted_sample_percent))

  }

  return(table_df)

}
